cpp/include/cuspatial/experimental/trajectory_bounding_boxes.cuh

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Compute the spatial bounding boxes of trajectories.
+ *
+ * Computes a tight bounding box around all points within each trajectory (points with the same ID).
+ *
+ * @note Assumes Object IDs and points are presorted by ID. This can be done using
+ * cuspatial::derive_trajectories.
+ *
+ * @tparam IdInputIt Iterator over object IDs. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-readable.
+ * @tparam PointInputIt Iterator over points. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-readable.
+ * @tparam PointOutputIt Iterator over output bounding box points. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-writeable.
+ *
+ * @param ids_first beginning of the range of input object ids
+ * @param ids_last end of the range of input object ids
+ * @param points_first beginning of the range of input point (x,y) coordinates
+ * @param bounding_box_minima_first beginning of the range of output minimum bounding box
+ *                                  coordinates, one per trajectory
+ * @param bounding_box_maxima_first beginning of the range of output maximum bounding box
+ *                                  coordinates, one per trajectory
+ * @param stream the CUDA stream on which to perform computations.
+ *
+ * @return A `std::pair` of iterators to the ends of the ranges of output minimum and maximum
+ *         bounding box coordinates.
+ */
+template <typename IdInputIt, typename PointInputIt, typename PointOutputIt>
+std::pair<PointOutputIt, PointOutputIt> trajectory_bounding_boxes(
+  IdInputIt ids_first,
+  IdInputIt ids_last,
+  PointInputIt points_first,
+  PointOutputIt bounding_box_minima_first,
+  PointOutputIt bounding_box_maxima_first,
```

Comment:
I wonder if we should let the algorithm only accept a single output iterator and leave user the burden to construct the zipped output iterator before they pass it in. I think it's more consistent with thrust's algorithm design.

Reply:
I looked at this and it's a bit tricky. This is a case where there are two outputs: the minimum and the maximum vec_2d point of each box. I can't think of a single example where Thrust expects a single zip iterator for algorithms that naturally produce more than one output.  Also, using this from the column-based API already requires zipping the min_x and min_y and zipping the  max_x and max_y columns together to output the 2D extrema. Zipping the extrema together would require the column-based API to do a 2-level zip. I also have to specify what type the extrema box is -- is it a pair? a tuple? or a new box struct?

I see three options.

1. Leave it as I have it now.
2. Do as you suggest, resulting in a 2-level zip in the column-based API, and single-level zips in header-only tests.
3. Introduce a new box struct type that is used in the header-only API. This would change the column-based API to use a 4-element zip (but single level).

I lean toward option 1 or 3. I don't like option 2.

Reply:
Michael convinced me to do #2...